File: tests/python/test_progression.py
# Gingo — Progression tests (Python)
# SPDX-License-Identifier: MIT

import pytest
from gingo import (
    Progression, Tree, Tradition, Schema,
    ProgressionMatch, ProgressionRoute,
    ScaleType, HarmonicFunction,
)


# ---------------------------------------------------------------------------
# Construction
# ---------------------------------------------------------------------------


def test_construction_from_scale_type():
    p = Progression("C", ScaleType.Major)
    assert p.tonic().natural() == "C"
    assert p.type() == ScaleType.Major


def test_construction_from_string():
    p = Progression("C", "major")
    assert p.tonic().natural() == "C"
    assert p.type() == ScaleType.Major


def test_construction_different_tonics():
    p1 = Progression("G", ScaleType.Major)
    assert p1.tonic().natural() == "G"

    p2 = Progression("A", ScaleType.NaturalMinor)
    assert p2.tonic().natural() == "A"
    assert p2.type() == ScaleType.NaturalMinor


def test_construction_minor_string():
    p = Progression("A", "natural minor")
    assert p.type() == ScaleType.NaturalMinor


# ---------------------------------------------------------------------------
# Traditions
# ---------------------------------------------------------------------------


def test_traditions_returns_list():
    traditions = Progression.traditions()
    assert isinstance(traditions, list)
    assert len(traditions) >= 2


def test_traditions_contains_harmonic_tree():
    traditions = Progression.traditions()
    names = [t.name for t in traditions]
    assert "harmonic_tree" in names


def test_traditions_contains_jazz():
    traditions = Progression.traditions()
    names = [t.name for t in traditions]
    assert "jazz" in names


def test_traditions_elements_are_tradition_objects():
    traditions = Progression.traditions()
    for t in traditions:
        assert isinstance(t, Tradition)


# ---------------------------------------------------------------------------
# Tree access
# ---------------------------------------------------------------------------


def test_tree_harmonic_tree():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    assert isinstance(t, Tree)
    branches = t.branches()
    assert len(branches) > 0
    assert "I" in branches


def test_tree_jazz():
    p = Progression("C", ScaleType.Major)
    t = p.tree("jazz")
    assert isinstance(t, Tree)
    branches = t.branches()
    assert len(branches) > 0
    assert "I" in branches
    assert "IIm" in branches
    assert "V7" in branches


def test_tree_invalid_tradition():
    p = Progression("C", ScaleType.Major)
    with pytest.raises(ValueError):
        p.tree("invalid")


# ---------------------------------------------------------------------------
# Tradition struct
# ---------------------------------------------------------------------------


def test_tradition_name_and_description():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    trad = t.tradition()
    assert isinstance(trad, Tradition)
    assert trad.name == "harmonic_tree"
    assert len(trad.description) > 0


def test_tradition_jazz_description():
    p = Progression("C", ScaleType.Major)
    t = p.tree("jazz")
    trad = t.tradition()
    assert trad.name == "jazz"
    assert len(trad.description) > 0


def test_tradition_to_dict():
    p = Progression("C", ScaleType.Major)
    trad = p.tree("harmonic_tree").tradition()
    d = trad.to_dict()
    assert isinstance(d, dict)
    assert set(d.keys()) == {"name", "description"}
    assert d["name"] == "harmonic_tree"
    assert isinstance(d["description"], str)
    assert len(d["description"]) > 0


# ---------------------------------------------------------------------------
# Schema struct
# ---------------------------------------------------------------------------


def test_schemas_harmonic_tree():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    schemas = t.schemas()
    assert len(schemas) > 0
    assert all(isinstance(s, Schema) for s in schemas)


def test_schemas_jazz():
    p = Progression("C", ScaleType.Major)
    t = p.tree("jazz")
    schemas = t.schemas()
    assert len(schemas) > 0

    names = [s.name for s in schemas]
    assert "ii-V-I" in names
    assert "turnaround" in names


def test_schema_attributes():
    p = Progression("C", ScaleType.Major)
    schemas = p.tree("jazz").schemas()
    ii_v_i = [s for s in schemas if s.name == "ii-V-I"][0]

    assert ii_v_i.name == "ii-V-I"
    assert len(ii_v_i.description) > 0
    assert ii_v_i.branches == ["IIm", "V7", "I"]


def test_schema_turnaround_branches():
    p = Progression("C", ScaleType.Major)
    schemas = p.tree("jazz").schemas()
    turnaround = [s for s in schemas if s.name == "turnaround"][0]

    assert turnaround.branches == ["I", "VIm", "IIm", "V7"]


def test_schema_to_dict():
    p = Progression("C", ScaleType.Major)
    schemas = p.tree("jazz").schemas()
    s = schemas[0]
    d = s.to_dict()
    assert isinstance(d, dict)
    assert set(d.keys()) == {"name", "description", "branches"}
    assert isinstance(d["name"], str)
    assert isinstance(d["description"], str)
    assert isinstance(d["branches"], list)


# ---------------------------------------------------------------------------
# Tree.is_valid()
# ---------------------------------------------------------------------------


def test_tree_is_valid_empty():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    assert t.is_valid([])


def test_tree_is_valid_single():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    assert t.is_valid(["I"])


def test_tree_is_valid_ii_v_i():
    p = Progression("C", ScaleType.Major)
    t = p.tree("jazz")
    valid = t.is_valid(["IIm", "V7", "I"])
    assert isinstance(valid, bool)


def test_tree_is_valid_invalid_branch():
    p = Progression("C", ScaleType.Major)
    t = p.tree("harmonic_tree")
    assert not t.is_valid(["I", "INVALID", "V7"])


# ---------------------------------------------------------------------------
# identify()
# ---------------------------------------------------------------------------


def test_identify_exact_ii_v_i():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["IIm", "V7", "I"])
    assert isinstance(m, ProgressionMatch)
    assert m.score > 0.0
    assert m.branches == ["IIm", "V7", "I"]


def test_identify_turnaround():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["I", "VIm", "IIm", "V7"])
    assert isinstance(m, ProgressionMatch)
    assert m.score > 0.0
    assert len(m.tradition) > 0


def test_identify_harmonic_tree_descending():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["I", "V7 / IIm", "IIm", "V7", "I"])
    assert isinstance(m, ProgressionMatch)
    assert m.score > 0.0
    assert m.tradition == "harmonic_tree"
    assert m.schema == "descending"


def test_identify_returns_best_match():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["I", "V7", "I"])
    assert isinstance(m, ProgressionMatch)
    assert m.score > 0.0
    assert isinstance(m.matched, int)
    assert isinstance(m.total, int)


def test_identify_empty_raises():
    p = Progression("C", ScaleType.Major)
    with pytest.raises(ValueError):
        p.identify([])


# ---------------------------------------------------------------------------
# deduce()
# ---------------------------------------------------------------------------


def test_deduce_returns_list():
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["I", "IIm"])
    assert isinstance(results, list)
    assert len(results) > 0


def test_deduce_elements_are_progression_match():
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["I", "IIm"])
    for r in results:
        assert isinstance(r, ProgressionMatch)


def test_deduce_sorted_by_score():
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["I", "IIm"])
    for i in range(len(results) - 1):
        assert results[i].score >= results[i + 1].score


def test_deduce_limit():
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["I", "IIm"], limit=2)
    assert len(results) <= 2


def test_deduce_partial_ii_v():
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["IIm", "V7"])
    assert len(results) > 0
    # Should find ii-V-I or related patterns
    has_jazz = any(r.tradition == "jazz" for r in results)
    assert has_jazz


def test_deduce_empty_raises():
    p = Progression("C", ScaleType.Major)
    with pytest.raises(ValueError):
        p.deduce([])


# ---------------------------------------------------------------------------
# predict()
# ---------------------------------------------------------------------------


def test_predict_returns_list():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    assert isinstance(results, list)
    assert len(results) > 0


def test_predict_elements_are_progression_route():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    for r in results:
        assert isinstance(r, ProgressionRoute)


def test_predict_sorted_by_confidence():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    for i in range(len(results) - 1):
        assert results[i].confidence >= results[i + 1].confidence


def test_predict_has_next_branch():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    for r in results:
        assert len(r.next) > 0


def test_predict_v7_after_iim():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["IIm"])
    # V7 should appear as a likely next branch from IIm
    nexts = [r.next for r in results]
    assert "V7" in nexts


def test_predict_with_tradition_filter():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"], tradition="jazz")
    assert len(results) > 0
    for r in results:
        assert r.tradition == "jazz"


def test_predict_path_extends_input():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    for r in results:
        assert len(r.path) == 3
        assert r.path[0] == "I"
        assert r.path[1] == "IIm"
        assert r.path[2] == r.next


def test_predict_empty_raises():
    p = Progression("C", ScaleType.Major)
    with pytest.raises(ValueError):
        p.predict([])


# ---------------------------------------------------------------------------
# ProgressionMatch.to_dict()
# ---------------------------------------------------------------------------


def test_progression_match_to_dict_keys():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["IIm", "V7", "I"])
    d = m.to_dict()
    assert isinstance(d, dict)
    expected = {"tradition", "schema", "score", "matched", "total", "branches"}
    assert set(d.keys()) == expected


def test_progression_match_to_dict_values():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["IIm", "V7", "I"])
    d = m.to_dict()
    assert isinstance(d["tradition"], str)
    assert isinstance(d["score"], float)
    assert isinstance(d["matched"], int)
    assert isinstance(d["total"], int)
    assert isinstance(d["branches"], list)
    assert d["branches"] == ["IIm", "V7", "I"]


# ---------------------------------------------------------------------------
# ProgressionRoute.to_dict()
# ---------------------------------------------------------------------------


def test_progression_route_to_dict_keys():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    assert len(results) > 0
    d = results[0].to_dict()
    assert isinstance(d, dict)
    expected = {"next", "tradition", "schema", "path", "confidence"}
    assert set(d.keys()) == expected


def test_progression_route_to_dict_values():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    d = results[0].to_dict()
    assert isinstance(d["next"], str)
    assert isinstance(d["tradition"], str)
    assert isinstance(d["schema"], str)
    assert isinstance(d["path"], list)
    assert isinstance(d["confidence"], float)


# ---------------------------------------------------------------------------
# repr() and str()
# ---------------------------------------------------------------------------


def test_progression_repr():
    p = Progression("C", ScaleType.Major)
    r = repr(p)
    assert len(r) > 0
    assert "C" in r


def test_progression_str():
    p = Progression("C", ScaleType.Major)
    s = str(p)
    assert len(s) > 0
    assert "C" in s


def test_tradition_repr():
    trad = Progression("C", ScaleType.Major).tree("jazz").tradition()
    r = repr(trad)
    assert "Tradition" in r
    assert "jazz" in r


def test_tradition_str():
    trad = Progression("C", ScaleType.Major).tree("jazz").tradition()
    s = str(trad)
    assert s == "jazz"


def test_schema_repr():
    schemas = Progression("C", ScaleType.Major).tree("jazz").schemas()
    r = repr(schemas[0])
    assert "Schema" in r


def test_schema_str():
    schemas = Progression("C", ScaleType.Major).tree("jazz").schemas()
    s = str(schemas[0])
    assert len(s) > 0


def test_progression_match_repr():
    p = Progression("C", ScaleType.Major)
    m = p.identify(["IIm", "V7", "I"])
    r = repr(m)
    assert "ProgressionMatch" in r


def test_progression_route_repr():
    p = Progression("C", ScaleType.Major)
    results = p.predict(["I", "IIm"])
    assert len(results) > 0
    r = repr(results[0])
    assert "ProgressionRoute" in r


# ---------------------------------------------------------------------------
# Cross-tradition coverage
# ---------------------------------------------------------------------------


def test_identify_across_traditions():
    """identify() should search all traditions and return the best match."""
    p = Progression("C", ScaleType.Major)

    # Jazz ii-V-I exact match
    m_jazz = p.identify(["IIm", "V7", "I"])
    assert m_jazz.tradition == "jazz"
    assert m_jazz.schema == "ii-V-I"
    assert m_jazz.score == 1.0

    # Harmonic tree descending
    m_ht = p.identify(["I", "V7 / IIm", "IIm", "V7", "I"])
    assert m_ht.tradition == "harmonic_tree"
    assert m_ht.schema == "descending"


def test_deduce_across_traditions():
    """deduce() should return matches from multiple traditions."""
    p = Progression("C", ScaleType.Major)
    results = p.deduce(["I", "IIm", "V7"])
    traditions = {r.tradition for r in results}
    # Should appear in at least one tradition
    assert len(traditions) >= 1


def test_minor_key_progression():
    """Progression should work with minor keys."""
    p = Progression("A", ScaleType.NaturalMinor)
    t = p.tree("harmonic_tree")
    branches = t.branches()
    assert "Im" in branches


def test_minor_key_jazz():
    p = Progression("A", ScaleType.NaturalMinor)
    t = p.tree("jazz")
    branches = t.branches()
    assert "Im" in branches
    assert "V7" in branches

File: tests/cpp/test_progression.cpp
// Gingo — Progression module tests
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/progression.hpp>
#include <gingo/tree.hpp>

#include <algorithm>
#include <string>
#include <vector>

using namespace gingo;

// ---------------------------------------------------------------------------
// Construction
// ---------------------------------------------------------------------------

TEST_CASE("Progression construction from ScaleType", "[progression]") {
    Progression p("C", ScaleType::Major);
    REQUIRE(p.tonic().natural() == "C");
    REQUIRE(p.type() == ScaleType::Major);
}

TEST_CASE("Progression construction from string", "[progression]") {
    Progression p("C", "major");
    REQUIRE(p.tonic().natural() == "C");
    REQUIRE(p.type() == ScaleType::Major);
}

TEST_CASE("Progression construction with different tonics", "[progression]") {
    Progression p1("G", ScaleType::Major);
    REQUIRE(p1.tonic().natural() == "G");

    Progression p2("A", ScaleType::NaturalMinor);
    REQUIRE(p2.tonic().natural() == "A");

    Progression p3("Bb", "major");
    REQUIRE(p3.tonic().semitone() == Note("Bb").semitone());
}

TEST_CASE("Progression construction with minor types", "[progression]") {
    Progression p1("D", ScaleType::NaturalMinor);
    REQUIRE(p1.type() == ScaleType::NaturalMinor);

    Progression p2("E", ScaleType::HarmonicMinor);
    REQUIRE(p2.type() == ScaleType::HarmonicMinor);

    Progression p3("F", ScaleType::MelodicMinor);
    REQUIRE(p3.type() == ScaleType::MelodicMinor);
}

// ---------------------------------------------------------------------------
// Traditions (static)
// ---------------------------------------------------------------------------

TEST_CASE("Progression traditions returns list with expected entries", "[progression]") {
    auto traditions = Progression::traditions();
    REQUIRE(!traditions.empty());

    std::vector<std::string> names;
    for (const auto& t : traditions) {
        names.push_back(t.name);
    }

    // Must contain both harmonic_tree and jazz
    REQUIRE(std::find(names.begin(), names.end(), "harmonic_tree") != names.end());
    REQUIRE(std::find(names.begin(), names.end(), "jazz") != names.end());
}

TEST_CASE("Progression traditions have descriptions", "[progression]") {
    auto traditions = Progression::traditions();
    for (const auto& t : traditions) {
        REQUIRE(!t.name.empty());
        REQUIRE(!t.description.empty());
    }
}

// ---------------------------------------------------------------------------
// Tree access
// ---------------------------------------------------------------------------

TEST_CASE("Progression tree harmonic_tree returns valid tree", "[progression]") {
    Progression p("C", ScaleType::Major);
    Tree t = p.tree("harmonic_tree");

    REQUIRE(t.tonic().natural() == "C");
    REQUIRE(t.type() == ScaleType::Major);
}

TEST_CASE("Progression tree harmonic_tree tradition metadata", "[progression]") {
    Progression p("C", ScaleType::Major);
    Tree t = p.tree("harmonic_tree");
    auto trad = t.tradition();

    REQUIRE(trad.name == "harmonic_tree");
    REQUIRE(!trad.description.empty());
}

TEST_CASE("Progression tree jazz returns valid tree", "[progression]") {
    Progression p("C", ScaleType::Major);
    Tree t = p.tree("jazz");

    REQUIRE(t.tonic().natural() == "C");
    REQUIRE(t.type() == ScaleType::Major);
}

TEST_CASE("Progression tree jazz tradition metadata", "[progression]") {
    Progression p("C", ScaleType::Major);
    Tree t = p.tree("jazz");
    auto trad = t.tradition();

    REQUIRE(trad.name == "jazz");
    REQUIRE(!trad.description.empty());
}

TEST_CASE("Progression tree jazz has branches", "[progression]") {
    Progression p("C", ScaleType::Major);
    Tree t = p.tree("jazz");
    auto branches = t.branches();

    REQUIRE(!branches.empty());
}

TEST_CASE("Progression tree invalid throws", "[progression]") {
    Progression p("C", ScaleType::Major);
    REQUIRE_THROWS_AS(p.tree("invalid"), std::invalid_argument);
}

TEST_CASE("Progression tree empty string throws", "[progression]") {
    Progression p("C", ScaleType::Major);
    REQUIRE_THROWS_AS(p.tree(""), std::invalid_argument);
}

// ---------------------------------------------------------------------------
// Tree tradition()
// ---------------------------------------------------------------------------

TEST_CASE("Tree tradition returns correct struct for harmonic_tree", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto trad = t.tradition();

    REQUIRE(trad.name == "harmonic_tree");
    REQUIRE(!trad.description.empty());
}

TEST_CASE("Tree tradition returns correct struct for jazz", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "jazz");
    auto trad = t.tradition();

    REQUIRE(trad.name == "jazz");
    REQUIRE(!trad.description.empty());
}

// ---------------------------------------------------------------------------
// Tree schemas()
// ---------------------------------------------------------------------------

TEST_CASE("Tree schemas for harmonic_tree are non-empty", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto schemas = t.schemas();

    REQUIRE(!schemas.empty());
}

TEST_CASE("Tree schemas for harmonic_tree contain expected patterns", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto schemas = t.schemas();

    std::vector<std::string> names;
    for (const auto& s : schemas) {
        names.push_back(s.name);
    }

    // Harmonic tree should include classic patterns
    bool has_descending = std::find(names.begin(), names.end(), "descending") != names.end();
    bool has_ascending  = std::find(names.begin(), names.end(), "ascending")  != names.end();
    bool has_direct     = std::find(names.begin(), names.end(), "direct")     != names.end();

    REQUIRE(has_descending);
    REQUIRE(has_ascending);
    REQUIRE(has_direct);
}

TEST_CASE("Tree schemas for jazz are non-empty", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "jazz");
    auto schemas = t.schemas();

    REQUIRE(!schemas.empty());
}

TEST_CASE("Tree schemas for jazz contain expected patterns", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "jazz");
    auto schemas = t.schemas();

    std::vector<std::string> names;
    for (const auto& s : schemas) {
        names.push_back(s.name);
    }

    // Jazz should include canonical patterns
    bool has_ii_V_I      = std::find(names.begin(), names.end(), "ii-V-I")      != names.end();
    bool has_turnaround  = std::find(names.begin(), names.end(), "turnaround")  != names.end();
    bool has_backdoor    = std::find(names.begin(), names.end(), "backdoor")    != names.end();

    REQUIRE(has_ii_V_I);
    REQUIRE(has_turnaround);
    REQUIRE(has_backdoor);
}

TEST_CASE("Tree schemas have valid structure", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto schemas = t.schemas();

    for (const auto& s : schemas) {
        REQUIRE(!s.name.empty());
        REQUIRE(!s.description.empty());
        REQUIRE(!s.branches.empty());
    }
}

TEST_CASE("Tree jazz schemas have valid structure", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "jazz");
    auto schemas = t.schemas();

    for (const auto& s : schemas) {
        REQUIRE(!s.name.empty());
        REQUIRE(!s.description.empty());
        REQUIRE(!s.branches.empty());
    }
}

// ---------------------------------------------------------------------------
// Tree is_valid (replaces is_valid_progression)
// ---------------------------------------------------------------------------

TEST_CASE("Tree is_valid single chord", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE(t.is_valid({"I"}));
}

TEST_CASE("Tree is_valid II-V-I", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    auto valid = t.is_valid({"IIm", "V7", "I"});
    // Classic II-V-I: check it doesn't crash and returns bool
    REQUIRE((valid == true || valid == false));
}

TEST_CASE("Tree is_valid empty progression", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE(t.is_valid({}));
}

TEST_CASE("Tree is_valid with invalid branch returns false", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "harmonic_tree");
    REQUIRE_FALSE(t.is_valid({"I", "INVALID", "V7"}));
}

TEST_CASE("Tree is_valid jazz tradition", "[progression][tree]") {
    Tree t("C", ScaleType::Major, "jazz");
    // Should work without crashing regardless of tradition
    auto valid = t.is_valid({"IIm", "V7", "I"});
    REQUIRE((valid == true || valid == false));
}

// ---------------------------------------------------------------------------
// Progression to_string
// ---------------------------------------------------------------------------

TEST_CASE("Progression to_string is non-empty", "[progression]") {
    Progression p("C", ScaleType::Major);
    auto str = p.to_string();

    REQUIRE(!str.empty());
}

TEST_CASE("Progression to_string includes tonic", "[progression]") {
    Progression p("C", ScaleType::Major);
    auto str = p.to_string();

    REQUIRE(str.find("C") != std::string::npos);
}

TEST_CASE("Progression to_string for different keys", "[progression]") {
    Progression p1("G", ScaleType::Major);
    REQUIRE(p1.to_string().find("G") != std::string::npos);

    Progression p2("A", ScaleType::NaturalMinor);
    REQUIRE(p2.to_string().find("A") != std::string::npos);
}

// ---------------------------------------------------------------------------
// Integration: Progression -> Tree round trip
// ---------------------------------------------------------------------------

TEST_CASE("Progression tree preserves tonic and type", "[progression][tree]") {
    Progression p("D", ScaleType::Major);

    Tree ht = p.tree("harmonic_tree");
    REQUIRE(ht.tonic().natural() == "D");
    REQUIRE(ht.type() == ScaleType::Major);

    Tree jz = p.tree("jazz");
    REQUIRE(jz.tonic().natural() == "D");
    REQUIRE(jz.type() == ScaleType::Major);
}

TEST_CASE("Progression tree for minor key", "[progression][tree]") {
    Progression p("A", ScaleType::NaturalMinor);
    Tree t = p.tree("harmonic_tree");

    REQUIRE(t.tonic().natural() == "A");
    REQUIRE(t.type() == ScaleType::NaturalMinor);

    auto branches = t.branches();
    REQUIRE(!branches.empty());
}
